### ISSUE TITLE
Feature Request: Bitwise Operations

### ISSUE THREAD

**Issue by avloss** [pytorch/pytorch#768]
it seems that there is SOME support for bitwise operations in NumPy:
https://docs.scipy.org/doc/numpy/reference/generated/numpy.bitwise_xor.html
Bitwise operations are also supported by CUDA:
https://docs.nvidia.com/cuda/cuda-c-programming-guide/#warp-shuffle-functions
yet, the implementation of binary operation `xor` in `PyTorch` is 
`return (self + other).eq(1)`
which takes two ByteTensor. My undertsanding is that this is inefficient, because only one Bit within every Byte of this tensor is actually encoding something. Please correct me if I'm wrong!!

for the record neither TensorFlow has bitwise operations

**Comment by fmassa** [comment]
There are some pending PRs implementing bitwise operations in the C backend https://github.com/torch/torch7/pull/933 and https://github.com/torch/cutorch/pull/699

**Comment by apaszke** [comment]
You can't optimize it out to work on a single bit, because loads from any memory (both CUDA and RAM) have a certain granularity, that is never smaller than a single byte anyway. Still, merging these two ops into a single kernel would make it faster, but I don't think it should be a bottleneck. Is it too slow for your use case? Or is it just a suggestion?

**Comment by fmassa** [comment]
@apaszke I think the PRs that I mentioned do what @avloss meant, as it uses the underlying C bitwise operators.

**Comment by apaszke** [comment]
@fmass I think so too. I'm just trying to decide on how to prioritize the task.

**Comment by avloss** [comment]
Thanks for quick response guys! @fmassa is right, I meant bitwise on bits within a byte - `11^21=30` (as is vanilla python). I imagine porting that from torch (given that PR is merged) is not a straight forward thing.

**Comment by fmassa** [comment]
@avloss I believe once the PRs in torch/cutorch are merged, integrating the changes in pytorch should not be difficult.

**Comment by soumith** [comment]
fixed via #1556

**Comment by manueldiaz96** [comment]
@soumith could you please link to how to use these bitwise operations?

Searching on the docs, they do not seem to be there:
https://pytorch.org/docs/master/search.html?q=bitwise&check_keywords=yes&area=default

Thanks!
